File: stack.kt
internal class Stack<T>(size: Int) {

    private var items: Array<Any?>
    private var index: Int

    internal init {
        this.items = arrayOfNulls<Any?>(size)
        this.index = 0
    }

    internal fun isEmpty(): Boolean {
        return this.index == 0
    }

    internal fun isFull(): Boolean {
        return this.index == this.items.size
    }

    internal fun peek(): T? {

        if (this.isEmpty()) {
            throw Exception("Cannot pop from an empty stack!")
        }

        return this.items[this.index - 1] as T?
    }

    internal fun pop(): T? {

        if (this.isEmpty()) {
            throw Exception("Cannot pop from an empty stack!")
        }

        this.index -= 1
        var item = this.items[this.index]
        this.items[this.index] = null
        return item as T?
    }

    internal fun push(item: T?): Unit {

        if (this.isFull()) {
            throw Exception("Cannot push into a full stack!")
        }

        this.items[this.index] = item
        this.index += 1
    }

    internal fun size(): Int {
        return this.index
    }

}

File: stack.go
// 1.9

type Stack struct {
    Items []int;
    Index int;
}

func NewStack(length int) *Stack {
    return &Stack{Items: [length]int, Index: 0}
}

func (s *Stack) Empty() bool {
    return s.Index == 0
}

func (s *Stack) Full() bool {
    return s.Index == len(s.Items)
}

func (s *Stack) Peek() (int, error) {

    if s.Empty() {
        return 0, errors.New("Cannot peek at an empty stack!")
    }

    return s.Items[s.Index], nil
}

func (s *Stack) Pop() (int, error) {

    if s.Empty() {
        return 0, errors.New("Cannot pop from an empty stack!")
    }

    s.Index -= 1
    i := s.Items[s.Index]
    s.Items[s.Index] = nil
    return i, nil
}

func (s *Stack) Push(i int) (error) {

    if s.Full() {
        errors.New("Cannot push into a full stack!")
    }

    s.Items = append(s.Items, i)
    s.Index += 1
    return nil
}

func (s *Stack) Size() int {
    return s.Index
}
